Update extractTextPlugin transformation

bitpshr opened the pull request:
**What kind of change does this PR introduce?**
Feature

**Did you add tests for your changes?**
Yes

**If relevant, did you update the documentation?**
N/A

**Summary**
After [some discussion](https://github.com/webpack/webpack-cli/issues/373) the decision was made to update our migration transformations to actively support migrating only from the previous version to the current version, which currently means `v3 - v4`. Migrating older versions such as `v1` or `v2` will no longer be supported. This PR begins this work by updating the `extractTextPlugin` transformation. Resolves #373.

_Note:_ `v3` configurations that contain multiple `ExtractTextPlugin` instances are not yet supported due to a [limitation](https://github.com/webpack-contrib/mini-css-extract-plugin/issues/45) with `MiniCssExtractPlugin`.

**Does this PR introduce a breaking change?**
Yes. This PR changes `ExtractTextPlugin` migration behavior and produces a `v4`-specific configuration.

**Other information**
This PR only includes updates to the `extractTextPlugin` transformation. The following issues have been opened for the remaining transformations:

- #382
- #383
- #384
- #385
- #386
- #387
- #388
- #389

bitpshr commented on lib/migrate/extractTextPlugin/__testfixtures__/extractTextPlugin.input.js:
```diff
@@ -1,16 +1,26 @@
-                use: ExtractTextPlugin.extract('style-loader', 'css-loader')
+                use: ExtractTextPlugin.extract({
+                    use: 'css-loader',
+                    fallback: ['style-loader', { loader: 'test-object-loader' }],
```
I tried to fully demonstrate that string, object, and array loader configurations all work via this test case.

bitpshr commented on lib/migrate/extractTextPlugin/extractTextPlugin.js:
```diff
@@ -1,58 +1,145 @@
+	// Add new loader
+	loaderPath && Object.assign(loaderPath, {
+		value: currentFallbackLoader ?
+			j.conditionalExpression(
```
`MiniCssExtractPlugin` does not support `fallback` and `disabled` options. Instead, users should manually swap out `MiniCssExtractPlugin.loader` with the loader that was normally used as a `fallback`, which is what we're doing with this ternary. More information can be found [here](https://github.com/webpack-contrib/mini-css-extract-plugin/issues/88#issuecomment-3786856530).

PlayMa256 commented:
I found some problems. I had the current configuration (that was not created by me and i think it is a good "point out of the curve", since people can still have things like this: https://gist.github.com/PlayMa256/0a51b41e1b0eba910869764f378bf932

This was the generated config: https://gist.github.com/PlayMa256/587b2d3890c9c96e3f3dbe567c8ecfce

* It has 2 empty variables, which were kept in the plugins list.
* `MiniCssPlugin` has null parameters. I think it is better to remove them other then having them null.
* Is there a way to remove the `extractCSS.extract` statements and replace with `MiniCssExtractPlugin.loader`?

bitpshr commented:
Good catch and feedback. On it now.

PlayMa256 commented:
Well, the null problems seems to be gone. In the example i sent you (which is kind absurd i admit), it still [have](https://gist.github.com/PlayMa256/db9dc93e2ae5d376e41e25cca5b108af) (`extractSCSS.extract...`). We can't control that much, neither handle the problem in this case, where there are `fallbacks` and such.

Could you create a test for multiple loaders in a rule?
I mean:
```
[
          'style-loader',
          'css-loader',
          ExtractTextPlugin...
]
```
A scenario similar to this? That could show problems if it fails.

bitpshr commented:
@PlayMa256 I updated the logic so options are only passed if they're defined and `null` is never passed. I also updated the tests to verify cases when `ExtractTextPlugin.extract` is an argument to a function call (like your example, `['css-hot-loader'].concat(ExtractTextPlugin.extract(...))`, and cases when `ExtractTextPlugin.extract` is a member of an array (like your multiple loaders scenario above.) All are working correctly.

However, after digging into multiple-instance support a bit more, I think we may need to only support migrating configurations with a single instance of `ExtractTextPlugin` until true multiple-instance support lands in `MiniCssExtractPlugin`. See webpack-contrib/mini-css-extract-plugin#45. Until that happens, it's not possible to translate per-instance options such as `filename`. We could technically translate `extractCss.extract`-type call statements to `ExtractTextPlugin.extract` statements, but it wouldn't accomplish much since `MiniCssExtractPlugin` only supports one definition in the `plugins` array for now.

PlayMa256 commented:
Oh, got it! I did not know about this problem, i just found this tricky `webpack` config and it could be a nice way to debug this feature :). No problem not supporting this, we could insert this disclaimer in the docs. I'm about to go home, as i arrive and have time to test, i can give you more feedback if necessary! Good job @billyjanitsch !!

ev1stensberg commented:
@ematipico this is a nice PR for you to test & to get more knowledge of the `migrate` internals. Easy testable as well.

bitpshr commented:
Thanks, I moved most functionality to `ast-utils` and made it generic so other transformations can use them going forward. This should be ready for another review when you have time @ev1stensberg @ematipico.

ev1stensberg commented on lib/migrate/extractTextPlugin/extractTextPlugin.js:
```diff
@@ -1,58 +1,111 @@
  * @returns {Node} ast - jscodeshift ast
  */
-
 module.exports = function(j, ast) {
```
named functions

ev1stensberg commented on lib/utils/ast-utils.js:
```diff
@@ -651,6 +651,155 @@ function createFunctionWithArguments(j, p, name) {
+ * @param {any} j — jscodeshift API
+ * @param {Node} node - Node to start search from
+ * @param {String} pluginName - name of the plugin to remove
+ * @param {Node | Void} - path to the root webpack configuration object if plugin is found
```
Missing return

ev1stensberg commented on lib/utils/ast-utils.js:
```diff
@@ -651,6 +651,155 @@ function createFunctionWithArguments(j, p, name) {
+ * Calls an function with a path wrapped in an array, using path.elements if available
+ *
+ * @param {NodePath} path - path to wrap in array
+ * @param {Function} action - called with array path
```
Missing return

bitpshr replied:
Added `@returns {Void}`

ev1stensberg commented on lib/migrate/extractTextPlugin/extractTextPlugin.js:
```diff
@@ -1,58 +1,111 @@
-				)
-			);
-			p.value.arguments = [newArgs];
+	const devLoaders = [];
```
Can you abstract all this into more functions that can be reused, like find value "filename" for a given prop

bitpshr replied:
I'm not sure I follow; we already abstracted functionality into simple, single-purpose functions like `getPluginOptionValue` and `getPluginLoaderValue`. If we start getting even more abstract than this, we risk targeting nodes that aren't actually the nodes we care about. For example, someone could define `filename` completely outside of the webpack configuration object, so we need to be very careful.

The functions I already pulled out (`createPluginByName`, `removePluginByName`, `getLoaderOptionValue`, `getPluginOptionValue`, and `removeRequire`) are reusable and safe.

bitpshr commented:
@ev1stensberg thanks for the reviews. Everything has been addressed except for the following comment:

> Could you extract these into more functions that could be reused?

What type of functions were you imagining? The functions I already added (`createPluginByName`, `removePluginByName`, `getLoaderOptionValue`, `getPluginOptionValue`, and `removeRequire`) are reusable and safe, and as abstract as any other function in `ast-utils`.

bitpshr commented:
Thanks @ematipico. Could you please see my [comment](https://github.com/webpack/webpack-cli/pull/390#issuecomment-381354455) above and take another look when you have a chance @ev1stensberg? Thanks in advance!

ev1stensberg commented on lib/migrate/extractTextPlugin/extractTextPlugin.js:
```diff
@@ -1,58 +1,111 @@
+	pluginExists && createPluginByName(j, ast, "MiniCssExtractPlugin", loaderOptions);
+
+	// Remove old loader
+	ast.find(j.Identifier, { name: "ExtractTextPlugin" })
```
This could be abstracted into an util, maybe you can make it general so that we can reuse it for later?

ev1stensberg commented on lib/migrate/extractTextPlugin/extractTextPlugin.js:
```diff
@@ -1,58 +1,111 @@
+	});
+
+	// Add new loader
+	const loader = currentFallbackLoader ?
```
This might be another util, `addLoader`

ev1stensberg commented on lib/migrate/extractTextPlugin/extractTextPlugin.js:
```diff
@@ -1,58 +1,111 @@
+
+	// Add new loader
+	const loader = currentFallbackLoader ?
+		j.conditionalExpression(
```
It might be too specific, is there another way of adding a loader?

bitpshr replied:
I created an `addConditionalLoader` utility function per your comment above. I think it'll be useful in the long run, as it makes it easy to add loaders based on `NODE_ENV`.

bitpshr commented:
Thanks @ev1stensberg. This should be as generic as possible now. Let me know your thoughts.

bitpshr commented:
Thanks @ev1stensberg. Rebased.
